dart/lib/src/http_client/failed_request_client.dart

```diff
@@ -0,0 +1,174 @@
+/// exception.
+/// This can be for example for the following reasons:
+/// - In an browser environment this can be requests which fail because of CORS.
+/// - In an mobile or desktop application this can be requests which failed
+///   because the connection was interrupted.
+///
+/// Additionally you can configre specific HTTP response codes to be consideres
+/// as a failed request. In the following example, the status codes 404 and 500
+/// are considered a failed request.
+///
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+///
+/// var client = FailedRequestClient(
+///   failedRequestStatusCodes: [404, 500]
+/// );
+/// ```
+///
+/// Remarks:
+/// If this client is used as a wrapper, a call to close also closes the
+/// given client.
+///
+/// The `FailedRequestClient` can be used as a standalone client like this:
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+///
+/// var client = FailedRequestClient();
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+///
+/// The `FailedRequestClient` can also be used as a wrapper for your own
+/// HTTP [Client](https://pub.dev/documentation/http/latest/http/Client-class.html):
+/// ```dart
+/// import 'package:sentry/sentry.dart';
+/// import 'package:http/http.dart' as http;
+///
+/// final myClient = http.Client();
+///
+/// var client = FailedRequestClient(client: myClient);
+/// try {
+///  var uriResponse = await client.post('https://example.com/whatsit/create',
+///      body: {'name': 'doodle', 'color': 'blue'});
+///  print(await client.get(uriResponse.bodyFields['uri']));
+/// } finally {
+///  client.close();
+/// }
+/// ```
+class FailedRequestClient extends BaseClient {
+  FailedRequestClient({
+    this.maxRequestBodySize = MaxRequestBodySize.small,
+    this.failedRequestStatusCodes = const [],
+    Client? client,
+    Hub? hub,
+  })  : _hub = hub ?? HubAdapter(),
+        _client = client ?? Client();
+
+  final Client _client;
+  final Hub _hub;
+
+  /// Configures up to which size request bodies should be included in events
+  final MaxRequestBodySize maxRequestBodySize;
+
+  /// Describes which HTTP status codes should be considered as a failed
+  /// requests.
+  ///
+  /// Per default no status code is considered a failed request.
+  final List<int> failedRequestStatusCodes;
+
+  @override
+  Future<StreamedResponse> send(BaseRequest request) async {
+    try {
+      final response = await _client.send(request);
+
+      if (failedRequestStatusCodes.contains(response.statusCode)) {
+        await _captureException(request: request);
+      }
+
+      return response;
+    } catch (exception, stackTrace) {
+      await _captureException(
+        exception: exception,
+        stackTrace: stackTrace,
+        request: request,
+      );
+
+      rethrow;
```

Comment:
There needs to be a good hint about this. If the user captures this himself he gets two crash reports.

Reply:
a way to solve this is via deduplication, see Java
https://github.com/getsentry/sentry-java/blob/b45944c93d53f68bc1e64f5ac4c9a5400b252d44/sentry/src/main/java/io/sentry/DuplicateEventDetectionEventProcessor.java

we can also make this in another PR btw if we see its necessary

Reply:
Ah nice

Reply:
Our event processor are just type aliases for methods though. So that would be another breaking change for v6

Reply:
lets check also if WeakRef exists and works for Web too

Reply:
If there's no WeakRef, I think we could just use a Queue which only keeps the last x events. Would that work?

Reply:
I created an issue for this: https://github.com/getsentry/sentry-dart/issues/480